File: python/hackerrank/practice/picking_numbers.py
"""
https://www.hackerrank.com/challenges/picking-numbers/problem
"""
n = int(input())
arr = list(map(int, input().split()))

max_value = 1
current_max = 1
arr = sorted(arr)
t = arr[0]
for i in range(1, len(arr)):
    if abs(arr[i] - t) <= 1:
        current_max += 1
    else:
        t = arr[i]
        current_max = 1
    max_value = max(max_value, current_max)

print(max_value)


# different approach
def trial():
    my_dict = dict()
    for i in arr:
        my_dict[i] = my_dict.setdefault(i, 0) + 1

    maxi = 0
    keys = sorted(my_dict.keys())

    maxi = my_dict[keys[0]]
    current_max = maxi
    for i in range(1, len(keys) - 1):
        print("Key:", i)
        if abs(keys[i] - keys[i - 1]) == 1:
            current_max += my_dict[keys[i]]
        else:
            current_max = my_dict[keys[i]]
        maxi = max(maxi, current_max)
    print(keys)
    print(my_dict)
    print(maxi)

File: Java/src/com/codex/hackerrank/PickingNumbers.java
package com.codex.hackerrank;

/*
 * https://www.hackerrank.com/challenges/picking-numbers/problem
 */
import java.util.Scanner;
import java.util.Set;
import java.util.TreeMap;

public class PickingNumbers {
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		TreeMap<Integer, Integer> map = new TreeMap<Integer, Integer>();

		for (int t = 0; t < n; t++) {
			int x = scanner.nextInt();
			if (map.containsKey(x)) {
				map.put(x, map.get(x) + 1);
			} else {
				map.put(x, 1);
			}
		}
		Set<Integer> set = map.keySet();
		Integer[] list = new Integer[n];
		Integer[] keys = set.toArray(list);
		int max = map.get(keys[0]);
		int currentMax = max;
		for (int i = 1; i < keys.length; i++) {
			if (Math.abs(keys[i] - keys[i + 1]) == 1) {
				currentMax += map.get(keys[i + 1]);
			} else {
				currentMax = map.get(keys[i]);
			}
			max = max(currentMax, max);
		}

		System.out.println(map);
		System.out.println(max);
		scanner.close();
	}

	private static int max(int a, int b) {
		return a > b ? a : b;
	}
}
